include/revng/Pipeline/LLVMContainer.h

```diff
@@ -0,0 +1,201 @@
+
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InspectableLLVMContainer.h"
+#include "revng/Pipeline/LLVMPipe.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/FunctionTags.h"
+
+namespace Pipeline {
+
+/// Base class that provides a reasonable
+/// default inplmentation for a llvm container
+class LLVMContainerImpl {
+protected:
+  std::unique_ptr<llvm::Module> Module;
+
+public:
+  LLVMContainerImpl(std::unique_ptr<llvm::Module> M) : Module(std::move(M)) {
+    revng_assert(Module != nullptr);
+  }
+
+  void mergeBackDerived(LLVMContainerImpl &ToMerge) {
+    llvm::Linker TheLinker(*Module);
+
+    bool Result = TheLinker.linkInModule(std::move(ToMerge.Module),
+                                         llvm::Linker::OverrideFromSrc);
+
+    revng_assert(!Result, "Linker failed");
+  }
+
+  const llvm::Module &getModule() const { return *Module; }
+  llvm::Module &getModule() { return *Module; }
+
+  llvm::Error storeToDisk(llvm::StringRef Path) const {
+
+    std::error_code EC;
+    llvm::raw_fd_ostream OS(Path, EC, llvm::sys::fs::F_None);
+    if (EC)
+      return llvm::createStringError(EC,
+                                     "Could not store to file module %s",
+                                     Path.str().c_str());
+    llvm::WriteBitcodeToFile(*Module, OS);
+    OS.flush();
+    return llvm::Error::success();
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef Path) {
+    if (not llvm::sys::fs::exists(Path)) {
+      Module = std::make_unique<llvm::Module>("revng.module",
+                                              Module->getContext());
+      return llvm::Error::success();
+    }
+
+    llvm::SMDiagnostic Error;
+    auto M = llvm::parseIRFile(Path, Error, Module->getContext());
+    if (!M)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "Could not parse file %s",
+                                     Path.str().c_str());
+
+    Module = std::move(M);
+    return llvm::Error::success();
+  }
+};
+
+template<typename T>
+void cloneMetadata(llvm::ValueToValueMapTy &Map, const T *From) {
+  llvm::SmallVector<std::pair<std::uint32_t, llvm::MDNode *>, 4> Metadata;
+
+  llvm::WeakTrackingVH Mapped = Map[From];
+  T &To = llvm::cast<T>(*Mapped);
+  From->getAllMetadata(Metadata);
+
+  for (const auto &[index, Node] : Metadata) {
+    auto *ToMap = llvm::dyn_cast<llvm::MDNode>(Node);
+    if (ToMap == nullptr)
+      continue;
+
+    auto MayeMappedNode = Map.getMappedMD(ToMap);
+    if (not MayeMappedNode.hasValue())
+      continue;
+
+    auto *ToSet = llvm::dyn_cast<llvm::MDNode>(*MayeMappedNode);
+    if (ToSet == nullptr)
+      continue;
+    To.setMetadata(index, ToSet);
+  }
+}
+
+/// Implementation that can be derived by anyone so that multiple identical
+/// llvmcontainers can exists so that inspector kinds do not pollute each other.
+template<typename MostDerived>
+class LLVMContainerBase : public InspectableContainer<MostDerived> {
```

Comment:
For coherence with base classes
```suggestion
template<typename Derived>
class LLVMContainerBase : public InspectableContainer<Derived> {
```